File: imogi_finance/imogi_finance/public/js/tax_invoice_fields.js
(() => {
  frappe.provide('imogi_finance.tax_invoice');

  const DEFAULT_FIELD_MAPS = {
    'Purchase Invoice': {
      fp_no: 'ti_fp_no',
      fp_date: 'ti_fp_date',
      npwp: 'ti_fp_npwp',
      dpp: 'ti_fp_dpp',
      ppn: 'ti_fp_ppn',
      ppnbm: 'ti_fp_ppnbm',
      ppn_type: 'ti_fp_ppn_type',
      status: 'ti_verification_status',
      notes: 'ti_verification_notes',
      duplicate_flag: 'ti_duplicate_flag',
      npwp_match: 'ti_npwp_match',
      ocr_status: 'ti_ocr_status',
      ocr_confidence: 'ti_ocr_confidence',
      ocr_raw_json: 'ti_ocr_raw_json',
      tax_invoice_pdf: 'ti_tax_invoice_pdf',
    },
    'Expense Request': {
      fp_no: 'ti_fp_no',
      fp_date: 'ti_fp_date',
      npwp: 'ti_fp_npwp',
      dpp: 'ti_fp_dpp',
      ppn: 'ti_fp_ppn',
      ppnbm: 'ti_fp_ppnbm',
      ppn_type: 'ti_fp_ppn_type',
      status: 'ti_verification_status',
      notes: 'ti_verification_notes',
      duplicate_flag: 'ti_duplicate_flag',
      npwp_match: 'ti_npwp_match',
      ocr_status: 'ti_ocr_status',
      ocr_confidence: 'ti_ocr_confidence',
      ocr_raw_json: 'ti_ocr_raw_json',
      tax_invoice_pdf: 'ti_tax_invoice_pdf',
    },
    'Sales Invoice': {
      fp_no: 'out_fp_no',
      fp_date: 'out_fp_date',
      npwp: 'out_fp_customer_npwp',
      dpp: 'out_fp_dpp',
      ppn: 'out_fp_ppn',
      ppnbm: 'out_fp_ppnbm',
      ppn_type: 'out_fp_ppn_type',
      status: 'out_fp_status',
      notes: 'out_fp_verification_notes',
      duplicate_flag: 'out_fp_duplicate_flag',
      npwp_match: 'out_fp_npwp_match',
      ocr_status: 'out_fp_ocr_status',
      ocr_confidence: 'out_fp_ocr_confidence',
      ocr_raw_json: 'out_fp_ocr_raw_json',
      tax_invoice_pdf: 'out_fp_tax_invoice_pdf',
    },
    'Tax Invoice OCR Upload': {
      fp_no: 'fp_no',
      fp_date: 'fp_date',
      npwp: 'npwp',
      dpp: 'dpp',
      ppn: 'ppn',
      ppnbm: 'ppnbm',
      ppn_type: 'ppn_type',
      status: 'verification_status',
      notes: 'verification_notes',
      duplicate_flag: 'duplicate_flag',
      npwp_match: 'npwp_match',
      ocr_status: 'ocr_status',
      ocr_confidence: 'ocr_confidence',
      ocr_raw_json: 'ocr_raw_json',
      tax_invoice_pdf: 'tax_invoice_pdf',
    },
  };

  const DEFAULT_COPY_KEYS = [
    'fp_no',
    'fp_date',
    'npwp',
    'dpp',
    'ppn',
    'ppnbm',
    'ppn_type',
    'status',
    'notes',
    'duplicate_flag',
    'npwp_match',
  ];

  const DATA = {
    fieldMaps: { ...DEFAULT_FIELD_MAPS },
    copyKeys: [...DEFAULT_COPY_KEYS],
  };

  const JSON_URL = '/assets/imogi_finance/json/tax_invoice_field_maps.json';
  const DEFAULT_FALLBACK = 'Purchase Invoice';

  const normalizeFieldMaps = (maybeMaps) => {
    if (!maybeMaps || typeof maybeMaps !== 'object') {
      return;
    }

    Object.entries(maybeMaps).forEach(([doctype, mapping]) => {
      if (!mapping || typeof mapping !== 'object') {
        return;
      }

      DATA.fieldMaps[doctype] = { ...mapping };
    });
  };

  const maybeLoadFromJson = async () => {
    try {
      const response = await fetch(JSON_URL, { credentials: 'same-origin' });
      if (!response.ok) {
        return;
      }

      const payload = await response.json();
      normalizeFieldMaps(payload?.field_maps || payload?.fieldMaps);
      if (Array.isArray(payload?.copy_keys || payload?.copyKeys)) {
        DATA.copyKeys = [...(payload.copy_keys || payload.copyKeys)];
      }
    } catch (error) {
      // eslint-disable-next-line no-console
      console.warn('Unable to load tax invoice field maps JSON', error);
    }
  };

  const getFieldMap = (doctype) => {
    const mapping = DATA.fieldMaps[doctype] || DATA.fieldMaps[DEFAULT_FALLBACK];
    return { ...mapping };
  };

  const getFieldMaps = () => ({ ...DATA.fieldMaps });

  const getCopyKeys = () => [...DATA.copyKeys];

  const getSharedCopyKeys = (sourceDoctype, targetDoctype) => {
    const sourceMap = getFieldMap(sourceDoctype);
    const targetMap = getFieldMap(targetDoctype);
    return getCopyKeys().filter((key) => sourceMap[key] && targetMap[key]);
  };

  imogi_finance.tax_invoice.getFieldMap = getFieldMap;
  imogi_finance.tax_invoice.getFieldMaps = getFieldMaps;
  imogi_finance.tax_invoice.getCopyKeys = getCopyKeys;
  imogi_finance.tax_invoice.getSharedCopyKeys = getSharedCopyKeys;

  // Fire-and-forget JSON refresh to keep JS maps aligned with Python.
  maybeLoadFromJson();
})(); 

File: imogi_finance/imogi_finance/tax_invoice_fields.py
from __future__ import annotations

import json
from copy import deepcopy
from pathlib import Path
from typing import Iterable, Mapping

DEFAULT_STANDARD_FIELD_MAP = {
    "fp_no": "ti_fp_no",
    "fp_date": "ti_fp_date",
    "npwp": "ti_fp_npwp",
    "harga_jual": "ti_fp_harga_jual",
    "potongan_harga": "ti_fp_potongan_harga",
    "dpp": "ti_fp_dpp",
    "ppn": "ti_fp_ppn",
    "ppnbm": "ti_fp_ppnbm",
    "ppn_type": "ti_fp_ppn_type",
    "status": "ti_verification_status",
    "notes": "ti_verification_notes",
    "duplicate_flag": "ti_duplicate_flag",
    "npwp_match": "ti_npwp_match",
    "ocr_status": "ti_ocr_status",
    "ocr_started_at": "ti_ocr_started_at",
    "ocr_confidence": "ti_ocr_confidence",
    "ocr_text": "ti_ocr_text",
    "ocr_raw_json": "ti_ocr_raw_json",
    "tax_invoice_pdf": "ti_tax_invoice_pdf",
}

DEFAULT_FIELD_MAP: dict[str, dict[str, str]] = {
    "Purchase Invoice": deepcopy(DEFAULT_STANDARD_FIELD_MAP),
    "Expense Request": deepcopy(DEFAULT_STANDARD_FIELD_MAP),
    "Sales Invoice": {
        "fp_no": "out_fp_no",
        "fp_date": "out_fp_date",
        "npwp": "out_fp_customer_npwp",
        "harga_jual": "out_fp_harga_jual",
        "potongan_harga": "out_fp_potongan_harga",
        "dpp": "out_fp_dpp",
        "ppn": "out_fp_ppn",
        "ppnbm": "out_fp_ppnbm",
        "ppn_type": "out_fp_ppn_type",
        "status": "out_fp_status",
        "notes": "out_fp_verification_notes",
        "duplicate_flag": "out_fp_duplicate_flag",
        "npwp_match": "out_fp_npwp_match",
        "ocr_status": "out_fp_ocr_status",
        "ocr_started_at": "out_fp_ocr_started_at",
        "ocr_confidence": "out_fp_ocr_confidence",
        "ocr_text": "out_fp_ocr_text",
        "ocr_raw_json": "out_fp_ocr_raw_json",
        "tax_invoice_pdf": "out_fp_tax_invoice_pdf",
    },
    "Tax Invoice OCR Upload": {
        "fp_no": "fp_no",
        "fp_date": "fp_date",
        "npwp": "npwp",
        "harga_jual": "harga_jual",
        "potongan_harga": "potongan_harga",
        "dpp": "dpp",
        "ppn": "ppn",
        "ppnbm": "ppnbm",
        "ppn_type": "ppn_type",
        "tax_rate": "tax_rate",  # 🔥 FIX: Add tax_rate mapping for dynamic PPN validation
        "status": "verification_status",
        "notes": "ocr_summary_json",  # 🔥 FIX: OCR JSON summary goes to separate field, not verification_notes
        "ocr_error_log": "ocr_error_log",  # 🔥 FIX: OCR technical errors go to dedicated field
        "duplicate_flag": "duplicate_flag",
        "npwp_match": "npwp_match",
        "ocr_status": "ocr_status",
        "ocr_started_at": "ocr_started_at",
        "ocr_confidence": "ocr_confidence",
        "ocr_text": "ocr_text",
        "ocr_raw_json": "ocr_raw_json",
        "tax_invoice_pdf": "tax_invoice_pdf",
    },
}

DEFAULT_COPY_KEYS: tuple[str, ...] = (
    "fp_no",
    "fp_date",
    "npwp",
    "harga_jual",
    "potongan_harga",
    "dpp",
    "ppn",
    "ppnbm",
    "ppn_type",
    "status",
    "notes",
    "duplicate_flag",
    "npwp_match",
)


def _load_field_map_data() -> tuple[dict[str, dict[str, str]], tuple[str, ...]]:
    json_path = Path(__file__).resolve().parent / "public" / "json" / "tax_invoice_field_maps.json"
    if not json_path.exists():
        return deepcopy(DEFAULT_FIELD_MAP), DEFAULT_COPY_KEYS

    try:
        data: Mapping[str, object] = json.loads(json_path.read_text())
        json_field_maps = data.get("field_maps") or {}
        json_copy_keys = data.get("copy_keys") or ()

        field_maps: dict[str, dict[str, str]] = deepcopy(DEFAULT_FIELD_MAP)
        if isinstance(json_field_maps, dict):
            for doctype, mapping in json_field_maps.items():
                if isinstance(mapping, dict):
                    field_maps[doctype] = mapping

        copy_keys: list[str] = list(DEFAULT_COPY_KEYS)
        if isinstance(json_copy_keys, (list, tuple)):
            copy_keys = [str(key) for key in json_copy_keys] or copy_keys

        return field_maps, tuple(copy_keys)
    except Exception:
        return deepcopy(DEFAULT_FIELD_MAP), DEFAULT_COPY_KEYS


FIELD_MAP, COPY_KEYS = _load_field_map_data()

UPLOAD_LINK_FIELDS: dict[str, str] = {
    "Purchase Invoice": "ti_tax_invoice_upload",
    "Expense Request": "ti_tax_invoice_upload",
    "Sales Invoice": "out_fp_tax_invoice_upload",
}

def get_field_map(doctype: str) -> dict[str, str]:
    return FIELD_MAP.get(doctype) or FIELD_MAP["Purchase Invoice"]


def get_field_maps() -> dict[str, dict[str, str]]:
    return deepcopy(FIELD_MAP)


def get_supported_doctypes() -> set[str]:
    return set(FIELD_MAP)


def get_upload_link_field(doctype: str) -> str | None:
    return UPLOAD_LINK_FIELDS.get(doctype)


def iter_copy_keys() -> Iterable[str]:
    return COPY_KEYS


def get_copy_keys() -> tuple[str, ...]:
    return COPY_KEYS


def get_tax_invoice_fields(doctype: str) -> set[str]:
    return set(get_field_map(doctype).values())
